File: app/src/main/java/com/beautifulquran/ui/reader/focus/ReaderFocusController.kt
package com.beautifulquran.ui.reader.focus

import androidx.compose.animation.core.FastOutSlowInEasing
import androidx.compose.animation.core.animate
import androidx.compose.animation.core.tween
import androidx.compose.foundation.gestures.animateScrollBy
import androidx.compose.foundation.gestures.scrollBy
import androidx.compose.foundation.lazy.LazyListState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.State
import androidx.compose.runtime.derivedStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.snapshotFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlin.math.abs

/**
 * The Compose-facing half of the focus engine: it holds the [LazyListState] and
 * the verse↔item lookups, and is the **single writer** to that scroll state.
 * Every programmatic scroll in the reader — jumping from the selector, following
 * the recitation (including the chapter-opening basmalah), returning to the
 * playing target, the initial "continue listening" settle — goes through
 * [focus], so nothing fights over the list.
 *
 * All the actual maths lives in the pure [FocusEngine]; this class only reads
 * `layoutInfo`, snapshots it, and hands it over.
 *
 * Created via [rememberReaderFocusController]; its per-composition inputs (the
 * lookups, the surah length, the top guard) are refreshed there each frame.
 */
class ReaderFocusController internal constructor(
    val listState: LazyListState,
) {
    /**
     * Focus-target key → index in the LazyColumn item list.
     * Ayah numbers are 1-based; [FocusEngine.CHAPTER_TOP_FOCUS_AYAH] (0) maps
     * to the dedicated basmalah list item on preface chapters.
     */
    internal var itemIndexOfAyah: Map<Int, Int> = emptyMap()

    /**
     * LazyColumn item index → ayah number (1-based), for ayah items only.
     * The basmalah item is intentionally omitted so the rail readout stays in
     * 1..N.
     */
    internal var ayahNumberByItemIndex: Map<Int, Int> = emptyMap()

    /** Highest verse number in the surah — the top of the rail's range. */
    internal var lastAyahNumber: Int = 1

    /**
     * Pixels of top chrome not already covered by the list's content padding.
     * Currently 0 — the app-bar/status strip is reserved through content
     * padding, so offset 0 is already clear of it — but kept as a knob for
     * future immersive layouts.
     */
    internal var topGuardPx: Int = 0

    /**
     * Pixels reserved at the bottom of the list viewport for the soft edge
     * fade / reading band above the player bar. Tall-verse detection and
     * word-band follow both treat this strip as outside the usable page.
     */
    internal var bottomGuardPx: Int = 0

    /**
     * Serializes every programmatic scroll. Selector jumps and recitation-follow
     * can fire from sibling `LaunchedEffect`s in the same frame; without this,
     * the second `focus()` cancels the first mid-slide via `MutatorMutex` and
     * the jump reads as a pop.
     */
    private val focusMutex = Mutex()

    /**
     * The reader's continuous position through the surah (ayah number + fraction),
     * for the rail marker. Derived from `layoutInfo`, so it recomputes only when
     * read and only while scrolling.
     */
    val focusedPosition: State<Float> = derivedStateOf { computeReadoutPosition() }

    /** The whole-number verse the reader is currently looking at. */
    val focusedAyah: State<Int> = derivedStateOf {
        focusedPosition.value.toInt().coerceIn(1, lastAyahNumber.coerceAtLeast(1))
    }

    /**
     * Where [ayahNumber] sits relative to its ideal focus position. Reads
     * `layoutInfo`, so call inside a `derivedStateOf` to stay reactive. Returns
     * an in-focus placement for a null/unknown verse (nothing to correct).
     * Pass [FocusEngine.CHAPTER_TOP_FOCUS_AYAH] for the basmalah list item.
     */
    fun placementOf(ayahNumber: Int?): FocusPlacement {
        val itemIndex = ayahNumber?.let { itemIndexOfAyah[it] }
            ?: return FocusPlacement(FocusZone.IN_FOCUS, 0)
        val info = listState.layoutInfo
        val viewportHeight = info.viewportSize.height
        if (viewportHeight <= 0) return FocusPlacement(FocusZone.IN_FOCUS, 0)
        val visible = info.visibleItemsInfo.firstOrNull { it.index == itemIndex }
        val geometry = if (visible != null) {
            TargetGeometry(
                topPx = visible.offset,
                heightPx = visible.size,
                isLaidOut = true,
                isAboveWhenOffscreen = false,
            )
        } else {
            TargetGeometry(
                topPx = 0,
                heightPx = 0,
                isLaidOut = false,
                isAboveWhenOffscreen = itemIndex < listState.firstVisibleItemIndex,
            )
        }
        return FocusEngine.placement(geometry, viewportHeight, topGuardPx)
    }

    /** True when [ayahNumber] is currently laid out taller than the *usable*
     *  viewport (full height minus top/bottom guards), so word-level following
     *  (not verse-level pinning) should drive its scroll. The bottom guard is
     *  the reading band above the player bar — without it, a nearly-full-height
     *  verse would keep word-follow off while its last lines sit under the fade. */
    fun exceedsViewport(ayahNumber: Int?): Boolean {
        val itemIndex = ayahNumber?.let { itemIndexOfAyah[it] } ?: return false
        val info = listState.layoutInfo
        val viewportHeight = info.viewportSize.height
        if (viewportHeight <= 0) return false
        val visible = info.visibleItemsInfo.firstOrNull { it.index == itemIndex } ?: return false
        val usable = (viewportHeight - topGuardPx - bottomGuardPx).coerceAtLeast(1)
        return visible.size > usable
    }

    /**
     * Secondary lyric constraint: scroll so the active word sits inside the
     * comfortable reading band. Serialized with [focus] so verse glides and
     * word follow never fight.
     *
     * [measureInViewport] is invoked **inside** the lock (and may run again
     * after a competing scroll finishes) so bounds stay live: top/bottom are
     * distances from the LazyColumn content-start edge (0 = top of the list
     * viewport). Return null when the word is not currently laid out.
     */
    suspend fun keepWordInView(
        bandTopMarginPx: Float,
        bandBottomMarginPx: Float,
        measureInViewport: () -> Pair<Float, Float>?,
    ) {
        focusMutex.withLock {
            val viewportHeight = listState.layoutInfo.viewportSize.height
            if (viewportHeight <= 0) return
            val bounds = measureInViewport() ?: return
            val delta = FocusEngine.wordBandDeltaPx(
                wordTopPx = bounds.first,
                wordBottomPx = bounds.second,
                viewportHeightPx = viewportHeight.toFloat(),
                topGuardPx = topGuardPx.toFloat(),
                bandTopMarginPx = bandTopMarginPx,
                bandBottomMarginPx = bandBottomMarginPx,
            )
            if (abs(delta) < 0.5f) return
            listState.animateScrollBy(
                delta,
                animationSpec = tween(
                    durationMillis = WORD_GLIDE_MS,
                    easing = FastOutSlowInEasing,
                ),
            )
        }
    }

    /**
     * Bring [ayahNumber] into focus — the one entry point for every programmatic
     * scroll. Waits for a real viewport before measuring (killing the
     * open-before-layout race), then lands the target on its adaptive anchor.
     * [animate] false snaps instantly (used for the initial settle).
     *
     * Pass [FocusEngine.CHAPTER_TOP_FOCUS_AYAH] (0) to home onto the basmalah
     * list item above ayah 1 — same glide / placement path as any verse.
     *
     * [preRoll] is for jumps the reader initiates by hand (selector, search,
     * return-to-verse). The pure [FocusEngine.planJump] decides the whole
     * trajectory:
     * - **Near** — one continuous home-scroll from here onto the verse.
     * - **Far** — teleport to a doorstep, then one continuous home-scroll across
     *   a distance-scaled residual (up to ~48 items / one full second) that
     *   re-aims at the live remaining distance every frame — smooth all the way
     *   until it lands exactly, with no rush-then-settle handoff stutter.
     *
     * Recitation-follow leaves [preRoll] off so lyric tracking stays a gentle
     * glide — still via the same continuous [animateHomeOnto] path (not a
     * one-shot measure), so the next verse across a page divider never pops
     * into place when it was not yet laid out. See [FocusEngine.planJump].
     */
    suspend fun focus(ayahNumber: Int, animate: Boolean, preRoll: Boolean = false) {
        focusMutex.withLock {
            focusLocked(ayahNumber, animate, preRoll)
        }
    }

    private suspend fun focusLocked(ayahNumber: Int, animate: Boolean, preRoll: Boolean) {
        val itemIndex = itemIndexOfAyah[ayahNumber] ?: return
        // Never measure against a zero viewport — wait for the first real layout.
        val viewportHeight = snapshotFlow { listState.layoutInfo.viewportSize.height }
            .first { it > 0 }

        val info = listState.layoutInfo
        val fromIndex = listState.firstVisibleItemIndex
        val visibleCount = info.visibleItemsInfo.size.coerceAtLeast(1)
        val totalCount = info.totalItemsCount.coerceAtLeast(1)

        if (preRoll) {
            val plan = FocusEngine.planJump(
                fromIndex = fromIndex,
                toIndex = itemIndex,
                visibleItemCount = visibleCount,
                totalItemCount = totalCount,
            )
            plan.doorstepIndex?.let { doorstep ->
                // Far: cover the bulk instantly so the home-scroll only has the
                // distance-scaled residual to show as travel.
                listState.scrollToItem(doorstep)
            }
            // One continuous decelerating home — no second settle phase.
            animateHomeOnto(itemIndex, viewportHeight, plan.durationMs)
            return
        }

        // Recitation-follow / non-preRoll: soft glide, with a simple doorstep
        // when the target is far beyond the measured window.
        if (FocusEngine.shouldTeleport(
                targetIndexDelta = itemIndex - fromIndex,
                visibleItemCount = visibleCount,
            )
        ) {
            val doorstep = if (itemIndex > fromIndex) {
                itemIndex - visibleCount
            } else {
                itemIndex + visibleCount
            }
            listState.scrollToItem(
                doorstep.coerceIn(0, (totalCount - 1).coerceAtLeast(0)),
            )
        }

        if (!animate) {
            // Instant settle (continue-listening open): force the item into
            // layout when needed, then snap by the exact remaining pixels.
            val delta = measureGlideDelta(itemIndex, viewportHeight) ?: return
            if (delta != 0) listState.scrollBy(delta.toFloat())
            return
        }

        // Continuous re-aim — same path as the hand-jump residual. Critical
        // across a mushaf page break: the next ayah often sits just past a
        // PageDivider and is not yet laid out, so a one-shot measure would
        // scrollToItem (a pop) before gliding. Estimating + re-aiming each
        // frame keeps the motion a smooth decelerating scroll onto the verse.
        animateHomeOnto(itemIndex, viewportHeight, GLIDE_MS)
    }

    /**
     * Continuously home onto [itemIndex]'s adaptive anchor over [durationMs].
     *
     * Runs as a single [LazyListState.scroll] so the decelerating motion holds
     * the MutatorMutex the whole way — no rush-then-settle handoff. Each frame
     * re-reads the live remaining distance (measured when the verse is laid
     * out, estimated from average item height when not) and scrolls the
     * FastOutSlowIn fraction of whatever is left — see
     * [FocusEngine.homeScrollStep]. Uneven ayah heights never produce an
     * undershoot that then gets corrected in a chunky second phase.
     */
    private suspend fun animateHomeOnto(
        itemIndex: Int,
        viewportHeight: Int,
        durationMs: Int,
    ) {
        if (abs(remainingPxToAnchor(itemIndex, viewportHeight)) < 0.5f) return

        listState.scroll {
            var lastProgress = 0f
            animate(
                initialValue = 0f,
                targetValue = 1f,
                animationSpec = tween(
                    durationMillis = durationMs.coerceAtLeast(1),
                    easing = FastOutSlowInEasing,
                ),
            ) { value, _ ->
                val remaining = remainingPxToAnchor(itemIndex, viewportHeight)
                val step = FocusEngine.homeScrollStep(remaining, value, lastProgress)
                lastProgress = value
                if (abs(step) >= 0.5f) {
                    scrollBy(step)
                }
            }
            // Consume any sub-pixel leftover from clamping — invisible as a jump.
            val leftover = remainingPxToAnchor(itemIndex, viewportHeight)
            if (abs(leftover) >= 0.5f) {
                scrollBy(leftover)
            }
        }
    }

    /**
     * Signed pixels still needed to put [itemIndex] on its adaptive anchor.
     * Uses real geometry when laid out; otherwise estimates from the average
     * laid-out item height (updated as new verses enter during the home-scroll).
     */
    private fun remainingPxToAnchor(itemIndex: Int, viewportHeight: Int): Float {
        val visible = listState.layoutInfo.visibleItemsInfo
            .firstOrNull { it.index == itemIndex }
        if (visible != null) {
            val anchor = FocusEngine.anchorOffsetPx(viewportHeight, topGuardPx, visible.size)
            return FocusEngine.glideDeltaPx(
                TargetGeometry(
                    visible.offset,
                    visible.size,
                    isLaidOut = true,
                    isAboveWhenOffscreen = false,
                ),
                anchor,
            ).toFloat()
        }
        val avg = averageLaidOutItemHeightPx().coerceAtLeast(1)
        val indexDelta = itemIndex - listState.firstVisibleItemIndex
        // Approximate the target's top as if every intervening item were `avg`
        // tall, accounting for the partial scroll of the first visible item.
        val approxTop = indexDelta * avg - listState.firstVisibleItemScrollOffset
        val approxAnchor = FocusEngine.anchorOffsetPx(viewportHeight, topGuardPx, avg)
        return (approxTop - approxAnchor).toFloat()
    }

    /** Mean height of currently laid-out items — used when the target is not
     *  yet measured. Recomputed each frame so the estimate improves as verses
     *  rush past during a far home-scroll. */
    private fun averageLaidOutItemHeightPx(): Int {
        val items = listState.layoutInfo.visibleItemsInfo
        if (items.isEmpty()) return 0
        return items.sumOf { it.size } / items.size
    }

    /**
     * Pixels to scroll so [itemIndex] lands on its adaptive anchor. Forces the
     * item into layout when it is not yet measured — used only by the instant
     * (`animate = false`) settle path, where a snap is intentional.
     */
    private suspend fun measureGlideDelta(itemIndex: Int, viewportHeight: Int): Int? {
        var target = listState.layoutInfo.visibleItemsInfo.firstOrNull { it.index == itemIndex }
        if (target == null) {
            listState.scrollToItem(itemIndex)
            target = listState.layoutInfo.visibleItemsInfo.firstOrNull { it.index == itemIndex }
        }
        val geometry = target?.let {
            TargetGeometry(it.offset, it.size, isLaidOut = true, isAboveWhenOffscreen = false)
        } ?: return null
        val anchor = FocusEngine.anchorOffsetPx(viewportHeight, topGuardPx, geometry.heightPx)
        return FocusEngine.glideDeltaPx(geometry, anchor)
    }

    private fun computeReadoutPosition(): Float {
        val info = listState.layoutInfo
        val viewportHeight = info.viewportSize.height
        val last = lastAyahNumber.coerceAtLeast(1)
        if (viewportHeight <= 0 || info.visibleItemsInfo.isEmpty()) return 1f

        val readingLine = FocusEngine.readingLinePx(viewportHeight, topGuardPx)
        val ayahItems = info.visibleItemsInfo.filter { ayahNumberByItemIndex.containsKey(it.index) }
        if (ayahItems.isEmpty()) return 1f
        // The verse crossing the reading line — not merely the first one peeking
        // over the top edge, so a 1px sliver never reads as a whole verse behind.
        val reading = ayahItems.lastOrNull { it.offset <= readingLine } ?: ayahItems.first()
        val readingAyah = ayahNumberByItemIndex[reading.index] ?: return 1f

        val lastItem = info.visibleItemsInfo.last()
        val tailVisible = lastItem.index == info.totalItemsCount - 1
        val snapshot = ReadoutSnapshot(
            readingAyah = readingAyah,
            readingAyahTopPx = reading.offset,
            readingAyahHeightPx = reading.size,
            readingLinePx = readingLine,
            tailVisible = tailVisible,
            tailBeyondFoldPx = if (tailVisible) {
                (lastItem.offset + lastItem.size - info.viewportEndOffset).coerceAtLeast(0)
            } else {
                0
            },
            tailHeightPx = lastItem.size,
            lastAyahNumber = last,
        )
        return FocusEngine.readoutPosition(snapshot)
    }

    private companion object {
        /** Soft recitation-follow / layout-reflow glide duration. Hand-initiated
         *  jumps use [FocusEngine.planJump]'s duration instead. */
        const val GLIDE_MS: Int = 700

        /** Word-band follow duration — short so line changes feel instant but
         *  still ease (matches the web port's WORD_GLIDE_MS). */
        const val WORD_GLIDE_MS: Int = 300
    }
}

/**
 * Remembers a [ReaderFocusController] bound to [listState] and refreshes its
 * per-composition inputs. The lookups are plain fields (not snapshot state):
 * they only change when the surah content changes, at which point the whole
 * screen recomposes and the derived states re-read them anyway.
 */
@Composable
fun rememberReaderFocusController(
    listState: LazyListState,
    itemIndexOfAyah: Map<Int, Int>,
    ayahNumberByItemIndex: Map<Int, Int>,
    lastAyahNumber: Int,
    topGuardPx: Int = 0,
    bottomGuardPx: Int = 0,
): ReaderFocusController {
    val controller = remember(listState) { ReaderFocusController(listState) }
    controller.itemIndexOfAyah = itemIndexOfAyah
    controller.ayahNumberByItemIndex = ayahNumberByItemIndex
    controller.lastAyahNumber = lastAyahNumber
    controller.topGuardPx = topGuardPx
    controller.bottomGuardPx = bottomGuardPx
    return controller
}

File: web/src/ui/reader/ReaderFocusController.ts
/**
 * DOM half of the focus engine — sole writer to the reader scroll container.
 * Pure maths stay in `engine/focus`; this class measures layout and scrolls.
 *
 * Port of Android `ui/reader/focus/ReaderFocusController.kt` for a non-virtual
 * scroll element (all ayah blocks are in the DOM).
 *
 * Verse advances and tall-verse line follow both use continuous `homeScrollStep`
 * re-aiming (not a one-shot delta or an instant `scrollTop` snap) so motion
 * stays butter-smooth onto the next verse or the next line.
 *
 * Progress timelines run through Motion with Material FastOutSlowIn — the same
 * curve Android uses via `FastOutSlowInEasing`.
 */
import { animate, type AnimationPlaybackControls } from 'motion'
import {
  FocusEngine,
  FocusZone,
  isChapterTopFocusTarget,
  type FocusPlacement,
  type TargetGeometry,
} from '../../engine/focus'
import { FAST_OUT_SLOW_IN } from '../motion/easing'

/**
 * Soft recitation-follow / verse-glide duration.
 * Slightly snappier than Android's 700 ms so next-verse feels responsive on
 * the web (especially mobile), while keeping the same FastOutSlowIn curve.
 */
const GLIDE_MS = 520
/** Tall-verse word-band follow — snappier than a full verse glide. */
const WORD_GLIDE_MS = 300
/**
 * Hand-initiated jump durations from FocusEngine are scaled down a touch so
 * scroll-to-verse feels faster without changing the pure engine constants.
 */
const JUMP_DURATION_SCALE = 0.82
const JUMP_DURATION_FLOOR_MS = 200
const ACTIVE_WORD_TOP_MARGIN_PX = 144
const ACTIVE_WORD_BOTTOM_MARGIN_PX = 132

export class ReaderFocusController {
  private scrollEl: HTMLElement | null = null
  private topGuardPx = 0
  private lastAyahNumber = 1
  private focusChain: Promise<void> = Promise.resolve()
  private focusEpoch = 0
  /** True while a Motion home-scroll is writing scrollTop. */
  private animating = false
  private activeControls: AnimationPlaybackControls | null = null

  bind(scrollEl: HTMLElement | null, lastAyahNumber: number, topGuardPx = 0) {
    this.scrollEl = scrollEl
    this.lastAyahNumber = Math.max(1, lastAyahNumber)
    this.topGuardPx = topGuardPx
  }

  /**
   * True while a programmatic glide is in flight. Reader scroll handlers
   * should skip heavy React readout updates during this window to avoid
   * per-frame re-renders (the main mobile lag source).
   */
  isAnimating(): boolean {
    return this.animating
  }

  /**
   * Bring [ayahNumber] onto its adaptive anchor. Concurrent calls serialize so
   * a selector jump and a recitation-follow never fight mid-glide.
   * Pass [FocusEngine.CHAPTER_TOP_FOCUS_AYAH] (0) to home onto the surah
   * header / basmalah preface above ayah 1.
   */
  focus(
    ayahNumber: number,
    options: { animate?: boolean; preRoll?: boolean } = {},
  ): Promise<void> {
    const animateScroll = options.animate !== false
    const preRoll = options.preRoll === true
    const epoch = ++this.focusEpoch
    this.focusChain = this.focusChain
      .catch(() => undefined)
      .then(() => this.focusLocked(ayahNumber, animateScroll, preRoll, epoch))
    return this.focusChain
  }

  /** Cancel any in-flight programmatic scroll (user grabbed the page). */
  cancel() {
    this.focusEpoch++
    this.activeControls?.stop()
    this.activeControls = null
    this.animating = false
  }

  placementOf(ayahNumber: number | null | undefined): FocusPlacement {
    if (ayahNumber == null) {
      return { zone: FocusZone.IN_FOCUS, distancePx: 0 }
    }
    const el = this.scrollEl
    if (!el) return { zone: FocusZone.IN_FOCUS, distancePx: 0 }
    const geom = this.geometryOf(ayahNumber)
    if (!geom) {
      return {
        zone: FocusZone.BELOW,
        distancePx: 0,
      }
    }
    return FocusEngine.placement(geom, el.clientHeight, this.topGuardPx)
  }

  exceedsViewport(ayahNumber: number | null | undefined): boolean {
    if (ayahNumber == null || !this.scrollEl) return false
    const target = this.ayahEl(ayahNumber)
    if (!target) return false
    const usable = Math.max(1, this.scrollEl.clientHeight - this.topGuardPx)
    return target.offsetHeight > usable
  }

  /** Continuous readout for the rail marker (ayah + fraction through it). */
  readoutPosition(): number {
    const el = this.scrollEl
    if (!el) return 1
    const viewport = el.clientHeight
    if (viewport <= 0) return 1
    const line = FocusEngine.readingLinePx(viewport, this.topGuardPx)
    const blocks = Array.from(el.querySelectorAll<HTMLElement>('.ayah-block'))
    if (blocks.length === 0) return 1

    const parentTop = el.getBoundingClientRect().top
    let reading: HTMLElement | null = null
    let readingTop = 0
    let readingHeight = 0
    // 1px slack: after home-scroll, the target can sit a subpixel below the
    // reading line; without slack, readout would stick on the previous ayah.
    const lineSlackPx = 1
    for (const block of blocks) {
      const top = block.getBoundingClientRect().top - parentTop
      if (top <= line + lineSlackPx) {
        reading = block
        readingTop = top
        readingHeight = block.offsetHeight
      }
    }
    if (!reading) {
      reading = blocks[0]!
      readingTop = reading.getBoundingClientRect().top - parentTop
      readingHeight = reading.offsetHeight
    }
    const readingAyah = Number(reading.dataset.ayah) || 1
    const last = blocks[blocks.length - 1]!
    const lastTop = last.getBoundingClientRect().top - parentTop
    const lastBottom = lastTop + last.offsetHeight
    const tailVisible = lastBottom <= viewport + 1 || lastTop < viewport
    const beyond = Math.max(0, lastBottom - viewport)

    return FocusEngine.readoutPosition({
      readingAyah,
      readingAyahTopPx: Math.round(readingTop),
      readingAyahHeightPx: Math.round(readingHeight),
      readingLinePx: line,
      tailVisible,
      tailBeyondFoldPx: Math.round(beyond),
      tailHeightPx: last.offsetHeight,
      lastAyahNumber: this.lastAyahNumber,
    })
  }

  focusedAyah(): number {
    return Math.min(
      this.lastAyahNumber,
      Math.max(1, Math.floor(this.readoutPosition())),
    )
  }

  /**
   * Secondary constraint: keep the active word inside the comfortable reading
   * band when the verse is taller than the viewport. Smooth continuous glide
   * (same `homeScrollStep` path as verse focus) — never an instant snap.
   * Serialized with [focus] so line follow and verse advances do not fight.
   */
  keepWordInView(wordEl: HTMLElement | null | undefined): void {
    if (!this.scrollEl || !wordEl) return
    const epoch = ++this.focusEpoch
    this.focusChain = this.focusChain
      .catch(() => undefined)
      .then(() => this.animateWordIntoBand(wordEl, WORD_GLIDE_MS, epoch))
  }

  private ayahEl(ayahNumber: number): HTMLElement | null {
    if (!this.scrollEl) return null
    if (isChapterTopFocusTarget(ayahNumber)) {
      return this.scrollEl.querySelector<HTMLElement>('#ayah-0, .basmalah-block')
    }
    return this.scrollEl.querySelector<HTMLElement>(`#ayah-${ayahNumber}`)
  }

  /** Focusable blocks in document order: optional basmalah, then ayahs. */
  private focusBlocks(): HTMLElement[] {
    const el = this.scrollEl
    if (!el) return []
    const blocks: HTMLElement[] = []
    const basmalah = el.querySelector<HTMLElement>('#ayah-0, .basmalah-block')
    if (basmalah) blocks.push(basmalah)
    blocks.push(...Array.from(el.querySelectorAll<HTMLElement>('.ayah-block')))
    return blocks
  }

  private blockFocusAyah(block: HTMLElement): number {
    if (
      block.id === 'ayah-0' ||
      block.classList.contains('basmalah-block') ||
      Number(block.dataset.ayah) === FocusEngine.CHAPTER_TOP_FOCUS_AYAH
    ) {
      return FocusEngine.CHAPTER_TOP_FOCUS_AYAH
    }
    return Number(block.dataset.ayah) || 1
  }

  private geometryOf(ayahNumber: number): TargetGeometry | null {
    const el = this.scrollEl
    const target = this.ayahEl(ayahNumber)
    if (!el || !target) return null
    return this.geometryOfBlock(target)
  }

  /**
   * Live viewport-relative geometry. Keep subpixel precision — rounding here
   * made home-scroll land a hair below the reading line, so readout then
   * reported the previous ayah after a rail jump.
   */
  private geometryOfBlock(target: HTMLElement): TargetGeometry | null {
    const el = this.scrollEl
    if (!el) return null
    const parentRect = el.getBoundingClientRect()
    const rect = target.getBoundingClientRect()
    return {
      topPx: rect.top - parentRect.top,
      heightPx: rect.height,
      isLaidOut: true,
      isAboveWhenOffscreen: false,
    }
  }

  private remainingPxToAnchor(ayahNumber: number): number {
    const el = this.scrollEl
    if (!el) return 0
    const geom = this.geometryOf(ayahNumber)
    if (!geom) return 0
    const anchor = FocusEngine.anchorOffsetPx(
      el.clientHeight,
      this.topGuardPx,
      geom.heightPx,
    )
    return FocusEngine.glideDeltaPx(geom, anchor)
  }

  /** Instantly place [block] on its adaptive anchor via live rects (not offsetTop). */
  private snapBlockOntoAnchor(block: HTMLElement) {
    const el = this.scrollEl
    if (!el) return
    const geom = this.geometryOfBlock(block)
    if (!geom) return
    const anchor = FocusEngine.anchorOffsetPx(
      el.clientHeight,
      this.topGuardPx,
      geom.heightPx,
    )
    const delta = FocusEngine.glideDeltaPx(geom, anchor)
    if (Math.abs(delta) >= 0.5) {
      el.scrollTop = Math.max(0, el.scrollTop + delta)
    }
  }

  private wordBandDelta(wordEl: HTMLElement): number {
    const el = this.scrollEl
    if (!el) return 0
    const parentRect = el.getBoundingClientRect()
    const rect = wordEl.getBoundingClientRect()
    return FocusEngine.wordBandDeltaPx(
      rect.top - parentRect.top,
      rect.bottom - parentRect.top,
      el.clientHeight,
      this.topGuardPx,
      ACTIVE_WORD_TOP_MARGIN_PX,
      ACTIVE_WORD_BOTTOM_MARGIN_PX,
    )
  }

  private jumpDurationMs(planMs: number): number {
    return Math.max(JUMP_DURATION_FLOOR_MS, Math.round(planMs * JUMP_DURATION_SCALE))
  }

  private async focusLocked(
    ayahNumber: number,
    shouldAnimate: boolean,
    preRoll: boolean,
    epoch: number,
  ): Promise<void> {
    const el = this.scrollEl
    if (!el || epoch !== this.focusEpoch) return
    const target = this.ayahEl(ayahNumber)
    if (!target) return

    const viewport = el.clientHeight
    if (viewport <= 0) return

    const blocks = this.focusBlocks()
    const toIndex = blocks.findIndex((b) => this.blockFocusAyah(b) === ayahNumber)
    if (toIndex < 0) return

    const line = FocusEngine.readingLinePx(viewport, this.topGuardPx)
    const parentTop = el.getBoundingClientRect().top
    let fromIndex = 0
    let best = Infinity
    blocks.forEach((b, i) => {
      const top = b.getBoundingClientRect().top - parentTop
      const dist = Math.abs(top - line)
      if (dist < best) {
        best = dist
        fromIndex = i
      }
    })

    const visibleCount = Math.max(
      1,
      blocks.filter((b) => {
        const top = b.getBoundingClientRect().top - parentTop
        const bottom = top + b.offsetHeight
        return bottom > 0 && top < viewport
      }).length,
    )

    if (preRoll) {
      const plan = FocusEngine.planJump(fromIndex, toIndex, visibleCount, blocks.length)
      if (plan.doorstepIndex != null && epoch === this.focusEpoch) {
        const door = blocks[plan.doorstepIndex]
        if (door) this.snapBlockOntoAnchor(door)
      }
      if (!shouldAnimate) {
        const delta = this.remainingPxToAnchor(ayahNumber)
        if (epoch === this.focusEpoch && Math.abs(delta) >= 0.5) {
          el.scrollTop = Math.max(0, el.scrollTop + delta)
        }
        return
      }
      await this.animateHomeOnto(ayahNumber, this.jumpDurationMs(plan.durationMs), epoch)
      return
    }

    if (
      FocusEngine.shouldTeleport(toIndex - fromIndex, visibleCount) &&
      epoch === this.focusEpoch
    ) {
      const doorstepIndex =
        toIndex > fromIndex
          ? Math.max(0, toIndex - visibleCount)
          : Math.min(blocks.length - 1, toIndex + visibleCount)
      const door = blocks[doorstepIndex]
      if (door) this.snapBlockOntoAnchor(door)
    }

    if (epoch !== this.focusEpoch) return
    if (!shouldAnimate) {
      const delta = this.remainingPxToAnchor(ayahNumber)
      el.scrollTop = Math.max(0, el.scrollTop + delta)
      return
    }
    // Recitation-follow: continuous re-aim (same path as hand-jump residual)
    // so the next verse lands smoothly even when intervening heights vary.
    await this.animateHomeOnto(ayahNumber, GLIDE_MS, epoch)
  }

  /**
   * Continuously home onto [ayahNumber]'s adaptive anchor over [durationMs].
   * Re-reads the live remaining distance every frame via [homeScrollStep].
   */
  private animateHomeOnto(
    ayahNumber: number,
    durationMs: number,
    epoch: number,
  ): Promise<void> {
    return this.animateRemaining(
      () => this.remainingPxToAnchor(ayahNumber),
      durationMs,
      epoch,
    )
  }

  /**
   * Continuously ease the active word into the reading band over [durationMs].
   */
  private animateWordIntoBand(
    wordEl: HTMLElement,
    durationMs: number,
    epoch: number,
  ): Promise<void> {
    if (!wordEl.isConnected) return Promise.resolve()
    return this.animateRemaining(() => this.wordBandDelta(wordEl), durationMs, epoch)
  }

  /**
   * Shared decelerating glide via Motion: each update scrolls the FastOutSlowIn
   * fraction of whatever distance [remainingPx] still reports.
   */
  private animateRemaining(
    remainingPx: () => number,
    durationMs: number,
    epoch: number,
  ): Promise<void> {
    const el = this.scrollEl
    if (!el) return Promise.resolve()
    if (Math.abs(remainingPx()) < 0.5) return Promise.resolve()

    this.activeControls?.stop()
    this.animating = true
    let lastProgress = 0

    return new Promise((resolve) => {
      const finish = () => {
        this.animating = false
        this.activeControls = null
        if (epoch === this.focusEpoch && this.scrollEl) {
          const leftover = remainingPx()
          if (Math.abs(leftover) >= 0.5) {
            this.scrollEl.scrollTop = Math.max(0, this.scrollEl.scrollTop + leftover)
          } else {
            // No leftover write — still notify listeners so rail readout
            // refreshes after we suppressed mid-glide scroll handlers.
            this.scrollEl.dispatchEvent(new Event('scroll'))
          }
        }
        resolve()
      }

      const controls = animate(0, 1, {
        duration: Math.max(0.001, durationMs / 1000),
        ease: [...FAST_OUT_SLOW_IN] as [number, number, number, number],
        onUpdate: (progress) => {
          if (epoch !== this.focusEpoch || !this.scrollEl) {
            controls.stop()
            this.animating = false
            this.activeControls = null
            resolve()
            return
          }
          const remaining = remainingPx()
          const step = FocusEngine.homeScrollStep(remaining, progress, lastProgress)
          lastProgress = progress
          if (Math.abs(step) >= 0.5) {
            this.scrollEl.scrollTop = Math.max(0, this.scrollEl.scrollTop + step)
          }
        },
        onComplete: finish,
      })
      this.activeControls = controls
    })
  }
}

export { ACTIVE_WORD_TOP_MARGIN_PX, ACTIVE_WORD_BOTTOM_MARGIN_PX, GLIDE_MS, WORD_GLIDE_MS }
